### ISSUE TITLE
Clarification 

### ISSUE THREAD

**Issue by jdakka** [hengma1001/CVAE_pilot_MD#1]
@hengma1001 It appears that you have the workflow coded start to finish with all the implementations in place based on Arvind's workflow [diagram] (https://www.dropbox.com/s/xdgyrkcjyx5rmv5/IBM_workflowForDeepMicroScope.pptx?dl=0), even if this implementation is a start 

Could you tell me at which point you are hurting? Are you lacking in scalability? Task coordination? Integration between HyperSpace and simulations? 

Answers would be helpful for me to understand how Cybertools can ameliorate certain pain points.

**Comment by hengma1001** [comment]
Hi, @jdakka 

Thank you for looking at the code. Sorry it's still very messy. 
At this point, the jobs for individual task in the workflow are asynchronous. I am not worrying about the scalability right now. 
Here are some issues that I had setting up the workflow. 
1. The current scheduler in the workflow is Celery with RabbitMQ. For every single time I need to train the autoencoder, some MD simulations have to be stopped to make the GPUs available. It will be sweet if there will be functions that can pause a job. 
2. I used to get the error from Celery workers, 'Connection reset by peer', especially when I was starting the autoencoder training right after stopping the MD jobs. This was resolved after I add `autoretry` in the task definition. You may have more experience dealing with the connection between worker and broker. 
3. Like I mentioned in _2_, the Celery was not very stable, and it failed sometime to return job status. I work it around by implementing the state into `class omm_job` and `class cvae_job`. 
4. The worker had to be started in bash or `subprocess` in python. It will be handy to have everything in python. But `subprocess` worked alright. 
5. This is a list of my responsibilities, posted it here for clarification and tracking jobs that need to be done. Feel free to let me know if you have any suggestion regarding these issues also. The Keras/Tensorflow duet is very easy to set up, but certain implementations could be a drawback in this workflow. 
    1. The most notable one is GPU memory issue. `K.clear_session()` and `del model` weren't able to release the GPU memory for further training or verification. _Fs-pep_ case was fine due to the small size, but the _P27_ case failed after ~50 iterations due to `ResourceExhaustedError`. 
    2. The `ExAB` case could not be trained due to the size of contact map, namely number of C_alpha.  

There are also some perks of Celery. 
1. The `flower` package can keep a visual track of the job status. 
2. Multiple cocurrent workers
3. There might be some others that I forgot. :bowtie:

Thank you again for checking the code. I hope the input will be helpful and let me know if you have any quesion.

**Comment by jdakka** [comment]
Hi @hengma1001 

Thanks for detailing your troubles.

I'll respond to them as I go but for now:

Regarding 1) It sounds like you're looking to checkpoint the simulations. Is it necessary to stop the simulations, while the autoencoders can begin training after the first 100K simulations frames? Can the autoencodes run using different workers on GPU resources that are not used by the simulations? 

Regarding 4) Is using processes in `python` a bottleneck? Our underlying stack is also implemented with `python` processes and threads. Do you have a failure rate of processes getting lost or not returning? We can compare failure modes between `celery` and `radical-pilot`. 

Regarding 5) This is exhausting the GPU memory and it sounds to me like you're asking for streaming capabilities. Currently we do not support streaming; we write everything to files. If a future task requires a file that was written by a previous task, we can create symbolic links for those pre-existing files, instead of copying. I have used `h5py` in the past to compress my data and re-create it in matrix format for convolutional neural nets using high resolution data. I also faced limitations with GPU memory. What worked for me was to build a "distributed" model that distributed each k-fold as a separate model. Essentially I had K instances of a model running, where each K is a K-fold. At the end, we took the average of all the k-folds (validation loss, test accuracy, etc.) Would it be possible to devise your autoencoder to use cross-validation? (It's been over a year since I've done ML so my question might be ill-posed). If not, I assume you also can't reduce the batch size...so we'll have to think of alternatives.

**Comment by hengma1001** [comment]
Hi, @jdakka 

For 1) I tried with not stopping the MD job while training the autoencoders. It is working and I may start to implement retraining autoencoder when the frame reaches a certain threshold. 

For 4) Both the MD code and autoencoder are using `python` as an API to the core in CUDA. Almost every time, it gave me the `error: [Errno 104] Connection reset by peer` error when I tried to start the job training the model. 

For 5) This is more like an issue with Keras/Tensorflow. 
    5.1> Celery only support simple streaming like a number or a string. I tried to pass some arrays from the worker to the main stream. It didn't work. Now most of the worker write information into `h5` file. 
    5.2> In each autoencoder task, I saved the weight instead of the model because the `load_model` cannot recognize the `lambda` function, and had to recreate the model and load the weight every time. 
    5.3> After training, the processes weren't free up the GPU memory as long as workers are still running. I got around it using `sys.exit()` to kill the task in the end. 
I am thinking to use `pytorch` for the autoencoder part as I discussed with Arvind. Hopefully that will resolve the issue.

**Comment by jdakka** [comment]
For 5.1) Have you considered using Dask? I remember reading that Dask has in-memory capabilities, i.e. like streaming, would this help to keep the contact maps in memory? I can look into it if you need.

**Comment by jdakka** [comment]
Hi @hengma1001 I've put together a rough outline of the steps for your algorithm (according to tasklets):

- **Tasks**: Simulation -- Simulation -- Simulation (Each on an independent GPU)
- **Task**: Collect trajectory info 
- **Task**: Create contact map of frames in .h5 format → input to CVAE
- **Tasks**: CVAE -- CVAE -- CVAE (Each on independent GPU, differ in hyper_dims) 
- **Task**: Collect weights
- **Task**: While True loop (for n_iterations):
      -- Check # of frames
      -- Perform inference for each model latent dimension (Check if this needs a GPU) → collect outliers
      -- Writes new pdb traj files
      -- Determine new outliers
      -- Calculate RMSD
      -- If conditions are met (len(traj) <10K and outliers present, and if GPU available): **Tasks**: (Simulations for outliers) 
      -- Increase the iteration counter

* If you could verify if anything is missing/incorrect that would be helpful for sketching a workflow diagram. 
* Question: the inference mode, does it require a GPU? I assumed no but wanted to check.

**Comment by hengma1001** [comment]
5.1> The h5 file should be good enough handling the streaming of data for now. I tried Dask in the beginning, but it was too rigid that all the frame work has to use the Dask array or dataframe. 

The outline looks right. 
When I was transferring the contact map to latent dimension, I used GPU to load the autoencoder model. Let me verify it though.

**Comment by jdakka** [comment]
Regarding the simulation tasks, you expressed that the celery workers' connections often time-out and tasks are lost as a result. I think it would be potentially useful to introduce the Ensemble Toolkit for managing the simulations, as this would allow you to spawn many concurrent simulations, without the issue of having to wait on GPUs to free up or losing tasks. 

Regarding the autoencoder training and inference, you wanted to test out `pytorch` to ensure that the GPU is freed once training is completed. 

Also, I'd like to verify the initial workflow [diagram](https://docs.google.com/drawings/d/1NjNfeQi1ypQ8eVDmS9t4X8S3-Wy4mzcBbcR76Qh-NcE/edit?usp=sharing ) as discussed above

**Comment by hengma1001** [comment]
I modified the OpenMM reporter, which can directly give you the contact map in 'h5' format, in addition to the MD trajectory. I was thinking about retraining the CVAE model after the simulation accumulates 1.6 times of the original number of frames.

Yes. Let me check out Ensemble Toolkit and pytorch.

**Comment by acadev** [comment]
Sorry to log in a little later. I think the diagram is a good representation of what you discussed @jdakka.  I would love to introduce the Ensemble toolkit to manage the jobs. We wanted something to run on our DGX=2 clusters so that they would run. Ultimately, Ensemble is what we should be using. 

Pytorch implementation is still under development. Should have some info later this month on the Pytorch version.

**Comment by jdakka** [comment]
Ok so we can pinpoint that the troubles in your workflow are with respect to Celery (connection issues and loss of tasks). I can help with the conversion of the current code with EnTK (Ensemble Toolkit). We can define an EnTK Task as a simulation job/executable with OpenMM or an analysis (CVAE/HyperSpace). Here is a reduced [version](https://github.com/radical-cybertools/htbac/blob/master/examples/openmm/openmm_example.py) that I did a while back that looks at creating one Task (OpenMM) and one Analysis Task (HyperSpace) . These Tasks belong to Stage 1 and Stage 2 respectively, where Stage 2 runs after Stage 1. This is what I had in mind, I'll also ask around to see if this is the best course of action with your code. Try to familiarize yourself with the PST (Pipeline, Stage, Task) model that Vivek gave an overview of a few months ago. It's straightforward, it's just to help with terminology. https://radicalentk.readthedocs.io/en/latest/entk.html#design

**Comment by acadev** [comment]
@jdakka: your understanding makes sense to me and I agree that the job essentially consists of 
Stage 1: one simulation task based on openMM and 
Stage 2: one analysis task based on Hyperspace. 

I would love to see if we can set this on EnTK, which I think is fairly straightforward. @hengma1001 can basically make them as modules (simulation task and the analysis task), which may be callable from EntK, is that right?

**Comment by acadev** [comment]
@jdakka : what version of OpenMM are you using? Also, do you know what version of Keras/ Tensorflow is available on Bridges?

**Comment by jdakka** [comment]
Which version of `OpenMM` I can use depends on you...On `Bridges` the following `CUDA` packages are available. 

`cuda/8.0   cuda/8.0RC cuda/9.0   cuda/9.0RC cuda/9.2` 

Let me know which version of `CUDA` and `OpenMM` you prefer. 

* keras:

keras/1.1.2                
keras/2.0.4                
keras/2.0.6_anaconda       
keras/2.0.6_anaconda_nogpu 
keras/2.2.0_tf1.7_py2_gpu  keras/2.2.0_tf1.7_py3_gpu

* Tensorflow: 

tensorflow/0.10.0               
tensorflow/0.12.1               
tensorflow/0.8.0                
tensorflow/1.0.1                
tensorflow/1.1.0                
tensorflow/1.7_py2_gpu
tensorflow/0.10.0_nogpu         
tensorflow/0.12.1_nogpu         
tensorflow/0.8.0_nogpu          
tensorflow/1.0.1_anaconda       
tensorflow/1.1.0_nogpu          
tensorflow/1.7_py2_nogpu
tensorflow/0.11.0               
tensorflow/0.7.1                
tensorflow/0.9.0                
tensorflow/1.0.1_anaconda_nogpu tensorflow/1.4_gpu              
tensorflow/1.7_py3_gpu
tensorflow/0.11.0_nogpu         
tensorflow/0.7.1_nogpu          
tensorflow/0.9.0_nogpu          
tensorflow/1.0.1_nogpu          
tensorflow/1.5_gpu              
tensorflow/1.8_py3_gpu

**Comment by hengma1001** [comment]
@jdakka Any version of `OpenMM` later than 7.1 should work. 
`Keras 2.2.0`
`TensorFlow 1.8.0`

From your list, I think you can use `keras/2.2.0_tf1.7_py2_gpu` and `tensorflow/1.7_py2_gpu`.

**Comment by jdakka** [comment]
Great! I've created the environment for `Bridges` 
Upon running `OpenMM` test installation on a P100 GPU I have these benchmarks: 

```
(cvae) [dakka@gpu045 dakka]$ python -m simtk.testInstallation
There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 1.97908e-05
Reference vs. CUDA: 2.15143e-05
CPU vs. CUDA: 1.55214e-05
Reference vs. OpenCL: 2.153e-05
CPU vs. OpenCL: 1.55379e-05
CUDA vs. OpenCL: 1.136e-07

```

**Comment by hengma1001** [comment]
Great. It looks like the OpenMM is installed successfully. 
You may also want to install `pytorch` and `torchvision` in the environment if it is not too much troublesome. I have a version of CVAE in pytorch working with the MNIST data so far.

**Comment by jdakka** [comment]
Done. I've tested this example with the environment: https://github.com/radical-cybertools/htbac/blob/master/examples/openmm/openmm_example.py. It works on Bridges-GPU nodes
